### DISCUSSION TITLE
SubscribeQuery - unexpected document path piece

### DISCUSSION THREAD

**Discussion by Autoneer** [SchneiderInfosystems/FB4D#133] [Q&A] [answered]
My subscribequery 

```
  cCollectionID := 'users';
  QueryStr := TStructuredQuery.CreateForCollection(cCollectionID,true);
  fConfig.Database.SubscribeQuery(QueryStr, OnChangedColDocument,
    OnDeletedColDocument,['users', fuid,'bookings','']);
```

returns the following message?

"targetChange": {    "targetChangeType": "REMOVE",    "targetIds": [      2    ],    "cause": {      "code": 3,      "message": "Unexpected document path piece: /bookings/"

However  this Get works 100%

```
    fconfig.Database.Get(['users', fuid,'bookings',''],nil,
      OnFirestoreGetList, OnFirestoreError);

```

**Comment by SchneiderInfosystems** [comment]
Check this solution:

```
  fDatabase.SubscribeQuery(TStructuredQuery.CreateForCollection('', true),
    OnFSChangedDocInCollection, OnFSDeletedDocCollection,
    ['users', fuid,'bookings']);
```
Thank you for your short feedback and a star, if you like this library.

**Reply by Autoneer** [reply]
No it did not fix the problem

```
  fConfig.Database.SubscribeQuery(TStructuredQuery.CreateForCollection('',true),
  OnChangedColDocument,  OnDeletedColDocument,
  ['users', fuid, 'bookings']);
```
Only difference between messages is the / after bookings  

"targetChange": { "targetChangeType": "REMOVE", "targetIds": [ 2 ], "cause": { "code": 3, "message": "Unexpected document path piece: /bookings"

**Reply by Autoneer** [reply]
This is the simple StartListener procedure

```
procedure TfmxMain.StartListener;
begin
  fConfig.Database.SubscribeQuery(TStructuredQuery.CreateForCollection('',true),
  OnChangedColDocument,  OnDeletedColDocument,
  ['users', fuid, 'bookings']);

  fConfig.Database.StartListener(OnStopListening, OnListenerError,
    OnAuthRevoked, OnConnectionStateChange);
end;

```

**Reply by SchneiderInfosystems** [reply]
I have analyzed the question even more deeply. Try the following

```
  fDatabase.SubscribeQuery(TStructuredQuery.CreateForCollection('bookings', true),
    OnFSChangedDocInCollection, OnFSDeletedDocCollection,
    ['users', fuid]);
```

I will add the ability to listen for sub-collections in the demo application.

**Reply by SchneiderInfosystems** [reply]
Check also the _Firestore Listener tab_ in the _DemoFB4D_ App from the newest commit for V1.3.1.438.

**Reply by Autoneer** [reply]
Fantastic thanks for the feedback Christopher,

It works if you set IncludesDescendants parameter to false in the StructuredQuery.

However for your information, if you set it to True as per your example I get the following message

"'all_descendants' in CollectionSelector with document parent is not supported by real-time queries."   

Thank you again it solved my problem :-)

**Comment by SchneiderInfosystems** [comment, marked as answer]
Correct, IncludesDescendants must be set to false.

Unfortunately, the _Firestore_ does not support listening to changes in nested collections.

To monitor all documents in the `users/<fuid>/bookings/*`subcollection, you need this query:

```
  fDatabase.SubscribeQuery(TStructuredQuery.CreateForCollection('bookings'),
    OnFSChangedDocInCollection, OnFSDeletedDocCollection, ['users', fuid]);
```
